src/bloqade/gemini/analysis/logical_validation/analysis.py

```diff
@@ -1,27 +1,50 @@
 from typing import Any
-from dataclasses import dataclass
+from dataclasses import field, dataclass
 
 from kirin import ir
+from kirin.interp import InterpreterError
 from kirin.lattice import EmptyLattice
 from kirin.analysis import Forward, ForwardFrame
 from kirin.validation import ValidationPass
 
 from bloqade import squin
+from bloqade.analysis.address import Address, AddressReg, AddressQubit, AddressAnalysis
 
 
+@dataclass
 class _GeminiLogicalValidationAnalysis(Forward[EmptyLattice]):
     keys = ["gemini.validate.logical"]
 
-    first_gate = True
     lattice = EmptyLattice
+    addr_frame: ForwardFrame[Address]
+
+    first_gates: dict[int, bool] = field(init=False, default_factory=dict)
 
     def eval_fallback(self, frame: ForwardFrame, node: ir.Statement):
         if isinstance(node, squin.gate.stmts.Gate):
-            # NOTE: to validate that only the first encountered gate can be non-Clifford, we need to track this here
-            self.first_gate = False
+            raise InterpreterError(f"Missing implementation for gate {node}")
 
         return tuple(self.lattice.bottom() for _ in range(len(node.results)))
 
+    def check_first_gate(self, qubits: ir.SSAValue) -> bool:
+        address = self.addr_frame.get(qubits)
+
+        if isinstance(address, AddressQubit):
+            is_first = self.first_gates.get(address.data, True)
+            self.first_gates[address.data] = False
+            return is_first
```

Comment:
based on the semantics this branch shuold never execute since it will always be an Ilist of qubits.